### PR TITLE
audit policy integration test with fake stackdriver

### LINKED ISSUE #25591: Extend action field in AuthorizationPolicy for audit
We propose to extend the "action" field in AuthorizationPolicy to support use cases like "audit" , in which case, the matching condition in AuthorizationPolicy can be used to determine when an audit log is generated. It will set a new boolean attribute e.g. 'istio.audit', which can be read by telemetry plugins to determine whether to log based on its value.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ X] Policies and Telemetry
[ X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

cc @louiscryan @smawson @liminw @mandarjog

### PR DESCRIPTION
**Pull request by xulingqing**
This PR tests audit policy with mock stackdriver in telemetry test framework.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[X] Policies and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+			for _, cl := range ctx.Clusters() {
+				scopes.Framework.Infof("Validating Audit policy and Telemetry for Cluster %v", cl)
+
+				args := map[string]string{
+					"Namespace": getEchoNamespaceInstance().Name(),
+				}
+				policies := tmpl.EvaluateAllOrFail(t, args, file.AsStringOrFail(t, "testdata/v1beta1-audit-authorization-policy.yaml.tmpl"))
+				ctx.Config().ApplyYAMLOrFail(t, "", policies...)
+				defer ctx.Config().DeleteYAMLOrFail(t, "", policies...)
```
Comment (gargnupur) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:44]: I think you should apply the policies before sending traffic.. this can actually go in setup itself?

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+				retry.UntilSuccessOrFail(t, func() error {
+					clName := cl.Name()
+					if err := validateMetrics(t, serverRequestCount, clientRequestCount, clName); err != nil {
+						return err
+					}
```
Comment (gargnupur) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:50]: I don't think, you need to validate metrics here..

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+					t.Logf("Metrics validated for cluster %v", clName)
+					if err := validateLogs(t, serverLogEntry, clName); err != nil {
```
Comment (gargnupur) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:52]: Should add validation for audit log.. this will validate generic server access log

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go (hunk 1 above)
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go]: Thanks for your advice.

File: tests/integration/telemetry/stackdriver/testdata/custom_bootstrap.yaml.tmpl [hunk 4]
```diff
@@ -30,7 +29,7 @@ data:
             "stackdriver_exporter_enabled": true,
             "stackdriver_grpc_service": {
               "google_grpc": {
-                "target_uri": "{{ .StackdriverAddress }}",
+                "target_uri": "cloudtrace.googleapis.com",
```
Comment (bianpengyuan) [inline tests/integration/telemetry/stackdriver/testdata/custom_bootstrap.yaml.tmpl:32]: Why are we testing with real sd backend?

Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/testdata/custom_bootstrap.yaml.tmpl]: I'm testing the audit function with real stackdriver and tried to debug if it works for request-header field matching of authz policy. But the current code does not work for request-header field matching, and I will need to enable rbac envoy filter for the integration test. Do you have any recommendation on how to enable rbac envoy filter debugging for integ test?

Comment (yangminzhu) [reply tests/integration/telemetry/stackdriver/testdata/custom_bootstrap.yaml.tmpl]: I think you can psss `--istio.test.kube.helm.values="global.proxy.componentLogLevel=rbac:debug"` to enable the component logging when running the test, it really doesn't make sense that the request-header matching not working while the path matching is working, we don't do anything special to request-header matching.

File: tests/integration/telemetry/stackdriver/testdata/server_audit_log.json.tmpl [hunk 5]
```diff
@@ -0,0 +1,24 @@
+{
+    "httpRequest": {
+      "requestMethod": "GET",
+      "requestUrl": "http://srv.{{ .EchoNamespace }}.svc.cluster.local/audit-all",
```
Comment (yangminzhu) [inline tests/integration/telemetry/stackdriver/testdata/server_audit_log.json.tmpl:4]: I think you should also check that `/audit-none` is not audit logged, and `/foo` and `/bar` is logged when header matched.

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+			sendTrafficForAudit(t)
+			t.Logf("Traffic sent to namespace %v", ns)
+			for _, cl := range ctx.Clusters() {
+				scopes.Framework.Infof("Validating Audit policy and Telemetry for Cluster %v", cl)
+
+				retry.UntilSuccessOrFail(t, func() error {
+					clName := cl.Name()
+					t.Logf("Collect Audit Log for cluster %v", clName)
+					if err := validateLogs(t, serverAuditLogEntry, clName); err != nil {
```
Comment (yangminzhu) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:55]: I'm very familiar with the telemetry test but does this make sure it is checking the audit log entry instead of the normal access log entry? @gargnupur  @bianpengyuan

Comment (gargnupur) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go]: The reason it should work is because audit log entry logs much less data then the normal access log entry, hence if it was validating normal access log entry, it will fail.

To make it full proof, if needed , we can add logName in verification.

Comment (yangminzhu) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go]: I see, I think we should verify the logName if it's not too hard to add, or at least put a comment about this if we don't do it now.

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+		newOptions(map[string][]string{"X-Header": {"bar"}}, "/bar"),
+		newOptions(map[string][]string{"X-Header": {"foo"}}, "/foo"),
+
+		// path "/audit-all" is matched in audit policy and all requests will be logged
+		newOptions(nil, "/audit-all"),
+		newOptions(map[string][]string{"X-Audit": {"foo"}}, "/audit-all"),
+		newOptions(map[string][]string{"X-Audit": {"bar"}}, "/audit-all"),
+	}
+
+	for _, cltInstance := range clt {
```
Comment (gargnupur) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:138]: you can just pass this as param to sendTraffic and not need to run the loop here

File: tests/integration/telemetry/stackdriver/stackdriver_filter_test.go [hunk 8]
```diff
@@ -96,28 +98,46 @@ func unmarshalFromTemplateFile(file string, out proto.Message, clName string) er
...
+						}
+						t.Logf("Traces validated")
+						if err := validateEdges(t, clName); err != nil {
+							return err
+						}
+						t.Logf("Edges validated")
+
+						return nil
+
+					}, retry.Delay(3*time.Second), retry.Timeout(80*time.Second))
```
Comment (gargnupur) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:130]: can you change this back to `retry.Delay(10*time.Second), retry.Timeout(2*time.Minute))` and see if it fixes your issue

Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:130]: Sure! this might be the reason it pass with GKE but failed when using Kind!

File: pkg/test/framework/features/allowlist.txt [hunk 9]
```diff
@@ -1514,6 +1515,7 @@ telemetry_outboundtrafficpolicy,TestOutboundTrafficPolicy_AllowAny
 telemetry_outboundtrafficpolicy,TestOutboundTrafficPolicy_RegistryOnly
 telemetry_requestclassification,TestRequestClassification
 telemetry_stackdriver,TestStackdriverMonitoring
+telemetry_stackdriver,TestStackdriverHTTPAuditLogging
```
Comment (howardjohn) [inline pkg/test/framework/features/allowlist.txt:1518]: I think this should be resolved by adding `.Features(...)` on the test instead of allowlisting it

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
```
Comment (howardjohn) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:136]: Aren't we sending n^2 requests? In this function, we loop over `clt`, and in the main function we loop over `clt` and call sendTrafficForAudit

Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:141]: Yes, it is. Are you suggesting any other sending traffic methods?

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:141]: Why do we need to send n^2 requests? there are a bunch of duplicated requests as far as I can tell?

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+				"Namespace": ns,
+			}
+			policies := tmpl.EvaluateAllOrFail(t, args, file.AsStringOrFail(t, auditPolicyForLogEntry))
+			ctx.Config().ApplyYAMLOrFail(t, ns, policies...)
+			defer ctx.Config().DeleteYAMLOrFail(t, ns, policies...)
+			t.Logf("Audit policy deployed to namespace %v", ns)
+
+			t.Logf("Traffic sent to namespace %v", ns)
+			for _, cltInstance := range clt {
+				scopes.Framework.Infof("Validating Audit policy and Telemetry for Cluster %v", cltInstance)
```
Comment (howardjohn) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:61]: nit: don't log plain `cltInstance`, it doesn't have a human readable format

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+				g.Go(func() error {
```
Comment (howardjohn) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:62]: This function will be called with the last cluster every time, see https://dinolai.com/notes/golang/golang-goroutines-in-loop.html

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go (hunk 11 above)
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:61]: Sure. Thanks!

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go (hunk 10 above)
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:141]: but actually there is only one client in the test

Comment (howardjohn) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:141]: The code is like this, as far as I can tell:
```go
for _, cltInstance := range clt {
  g.Go(func() {
     // sendTraffic 
     for _, cltInstance := range clt {
        echo.Call{..}
     }
  })
}
```

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go (hunk 12 above)
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:63]: I removed the loop in sendTraffic. Thanks!

File: pkg/test/framework/features/allowlist.txt (hunk 9 above)
Comment (xulingqing) [reply pkg/test/framework/features/allowlist.txt]: Do you think this is correct https://github.com/istio/istio/pull/28680/files#diff-012ed15b784ec47d12f52191781b2ce89c73d01996c592067bbeef9b3d4d250eR47 ? is there any other config required for the Features label?

File: pkg/test/framework/components/stackdriver/kube.go [hunk 13]
```diff
@@ -174,6 +174,8 @@ func (c *kubeComponent) ListLogEntries() ([]*loggingpb.LogEntry, error) {
 	for _, l := range r.Entries {
 		// Remove fields that do not need verification
 		l.Timestamp = nil
+		l.Trace = ""
```
Comment (bianpengyuan) [inline pkg/test/framework/components/stackdriver/kube.go:177]: I agree that it is not necessary to verify trace and span id, but we do verify log entry before this change. Wondering what is in this change that makes this issue surface?

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl [hunk 14]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (bianpengyuan) [inline tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl:1]: How can I tell if this is a audit log entry or normal access log? I think we'd need to verify the name the log entry?

File: tests/integration/telemetry/stackdriver/stackdriver_filter_test.go [hunk 15]
```diff
@@ -95,29 +97,48 @@ func unmarshalFromTemplateFile(file string, out proto.Message, clName string) er
 // TestStackdriverMonitoring verifies that stackdriver WASM filter exports metrics with expected labels.
 func TestStackdriverMonitoring(t *testing.T) {
 	framework.NewTest(t).
+		Features("observability.telemetry.stackdriver").
```
[the remark this answers is not in the document]
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:100]: Thanks!

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl (hunk 14 above)
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl:1]: audit log have less labels than access log. Now audit log do not support add a name for it.

File: pkg/test/framework/components/stackdriver/kube.go (hunk 13 above)
Comment (xulingqing) [reply pkg/test/framework/components/stackdriver/kube.go:177]: I think it might because accesslog verify trace and edge with validateTrace and ValidateEdge function. only validate log are using this ListLogEntries() func.

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+						errAuditNone := validateLogs(t, serverAuditNoneLogEntry, clName)
+						if errAuditNone == nil {
+							t.Logf("None Audit Log validated for cluster %v", clName)
+						}
+						if errAuditFoo == nil && errAuditBar == nil && errAuditAll == nil && errAuditNone != nil {
+							return nil
+						}
+
+						var errs []string
+						errs = append(errs, errAuditAll.Error(), errAuditFoo.Error(), errAuditBar.Error(), errAuditNone.Error())
```
Comment (yangminzhu) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:94]: You can create the `errs` before the first validateLogs() and append to it if the error is not nil. Then you can just check len(errs) != 0 in the end instead of specifying errAuditAll, errAuditFoo, errAuditBar and errAuditNone separately.

File: tests/integration/telemetry/stackdriver/testdata/traffic_assertion.json.tmpl [hunk 17]
```diff
@@ -13,7 +13,7 @@
         "canonical_service": "srv",
         "canonical_revision": "v1"
     },
-    "protocol": "PROTOCOL_GRPC",
+    "protocol": "PROTOCOL_HTTP",
```
Comment (yangminzhu) [inline tests/integration/telemetry/stackdriver/testdata/traffic_assertion.json.tmpl:16]: Why is this changed?

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl [hunk 18]
```diff
@@ -0,0 +1,25 @@
+{
```
[the remark this answers is not in the document]
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl:1]: there should not be log with /audit-none path, so in the check auditNone err should not be nil

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go (hunk 16 above)
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:94]: Sure! Thanks!

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl (hunk 14 above)
Comment (bianpengyuan) [reply tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl:1]: No audit log is named differently. See https://github.com/istio/proxy/blob/45ef14104ad84554f3600e5340dd59d9764a043c/extensions/stackdriver/log/logger_test.cc#L138 for an example full audit log message, which is named as `projects/test_project/logs/server-istio-audit-log`, whereas normal access log is named as `server-accesslog-stackdriver`.

I think what you should do is 1) extend [fake stackdriver](https://github.com/istio/proxy/blob/45ef14104ad84554f3600e5340dd59d9764a043c/test/envoye2e/stackdriver_plugin/fake_stackdriver.go#L141) to support distinguish different log types and returns audit log entries on a different REST URL (currently all log entries are returned by [/logentries](https://github.com/istio/proxy/blob/45ef14104ad84554f3600e5340dd59d9764a043c/test/envoye2e/stackdriver_plugin/fake_stackdriver.go#L486). we should create a new url like `/auditlogentries`) 2) then update the test here to update the listLogEntry method to list audit log entry only.

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl (hunk 18 above)
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl:1]: The inconsistency of validation condition is confusing. You should not verify the full log entry, but instead just go through the audit log entries and make sure there is no entry with path `audit-none`.

File: pkg/test/framework/components/stackdriver/kube.go (hunk 13 above)
Comment (bianpengyuan) [reply pkg/test/framework/components/stackdriver/kube.go:177]: Actually no I think it might because of trace sampling. Anyway, I think it is right to not verify trace related labels.

File: tests/integration/telemetry/stackdriver/testdata/traffic_assertion.json.tmpl (hunk 17 above)
Comment (bianpengyuan) [reply tests/integration/telemetry/stackdriver/testdata/traffic_assertion.json.tmpl:16]: +1

Comment (douglas-reid) [reply tests/integration/telemetry/stackdriver/testdata/traffic_assertion.json.tmpl]: edge reporting is cached and keyed by peers. protocol is not part of the cache key. so, if whatever test ordering happens in the environment with this test causes HTTP traffic to be sent first in some cases.

i'd suggest "zero"-ing out the protocol bits in the validation, so that there is a match regardless of actual protocol. 

we could "fix" edge reporting, but it is going away within a release cycle most likely, so we can be blunt about the fix.

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl (hunk 14 above)
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl:1]: Yeah thanks for showing also this great approach and elaborating on the details. Since my approach is also workable and trying to catch the 1.8 release. Maybe we can merge this one and then refactor code shortly.

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl (hunk 18 above)
[the remark this answers is not in the document]
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl:1]: @gargnupur Hi Nupur, is there a way to filter out audit log with label when testing with fack stackdriver? Thanks!

File: tests/integration/telemetry/stackdriver/stackdriver_filter_test.go [hunk 19]
```diff
@@ -162,6 +167,8 @@ values:
 	cfg.Values["telemetry.v2.stackdriver.enabled"] = "true"
 	cfg.Values["telemetry.v2.stackdriver.logging"] = "true"
 	cfg.Values["telemetry.v2.stackdriver.topology"] = "true"
+	cfg.Values["telemetry.v2.stackdriver.configOverride.enable_audit_log"] = "true"
+	cfg.Values["global.proxy.componentLogLevel"] = "rbac:debug"
```
Comment (gargnupur) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_test.go:171]: Is this needed? I thought this is just for debugging purposes?

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl (hunk 18 above)
[the remark this answers is not in the document]
Comment (gargnupur) [reply tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl:1]: Agreed with Peter, but problem is that you will find `audit-none` server access log without the refactoring Peter mentioned above...

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl (hunk 14 above)
Comment (gargnupur) [reply tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_all_log.json.tmpl:1]: We can also add logname from WriteLogEntriesRequest to logEntry and then filter here in the test.. that might be easiest.. @bianpengyuan : wdyt? 

Definitely, something to follow-up quickly if this PR needs to be submitted sooner..

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+						}
+
+						errAuditAll := validateLogs(t, serverAuditAllLogEntry, clName, stackdriver.ServerAuditLog)
+						if errAuditAll == nil {
+							t.Logf("All Audit Log validated for cluster %v", clName)
+						} else {
+							errs = append(errs, errAuditAll.Error())
+						}
+
+						errAuditNone := validateLogs(t, serverAuditNoneLogEntry, clName, stackdriver.ServerAuditLog)
```
Comment (bianpengyuan) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:97]: Please update the none audit log verification to verify there is no audit log entry with `/audit-none` as log path, and get rid of the `server_audit_none_log.json.tmpl` file. It could simply be:
```go
entries, err := sdInst.ListLogEntries(stackdriver.ServerAuditLog)
if err != nil {
  errs = append(errs, err)
} else {
  for _, l := range entries {
    if l.httpRequest != nil && strings.HasSuffix(l.httpReuqest.requestUrl, "audit-none") {
      errs = append(errs, errors.New("unwanted audit log entry `/audit-none` recieved."))
    }
  }
}
```

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+			ns := getEchoNamespaceInstance().Name()
+			args := map[string]string{
+				"Namespace": ns,
+			}
+			policies := tmpl.EvaluateAllOrFail(t, args, file.AsStringOrFail(t, auditPolicyForLogEntry))
+			ctx.Config().ApplyYAMLOrFail(t, ns, policies...)
+			defer ctx.Config().DeleteYAMLOrFail(t, ns, policies...)
+			t.Logf("Audit policy deployed to namespace %v", ns)
+
+			t.Logf("Traffic sent to namespace %v", ns)
```
Comment (bianpengyuan) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:62]: Should this log line move to line 70? Or move into `sendTrafficForAudit` fucntion?

File: pkg/test/framework/components/stackdriver/kube.go [hunk 22]
```diff
@@ -172,29 +188,34 @@ func (c *kubeComponent) ListLogEntries() ([]*loggingpb.LogEntry, error) {
 	}
 	var ret []*loggingpb.LogEntry
 	for _, l := range r.Entries {
```
[the remark this answers is not in the document]
Comment (xulingqing) [reply pkg/test/framework/components/stackdriver/kube.go:190]: Thanks a lot!

File: tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl (hunk 18 above)
[the remark this answers is not in the document]
Comment (xulingqing) [reply tests/integration/telemetry/stackdriver/testdata/security_authz_audit/server_audit_none_log.json.tmpl]: I rewrite the audit-none validation logic in the PR. Thanks a lot for all suggestion and instruction!

**Review by bianpengyuan — APPROVED** [review]
thanks!

File: tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+			args := map[string]string{
+				"Namespace": ns,
+			}
+			policies := tmpl.EvaluateAllOrFail(t, args, file.AsStringOrFail(t, auditPolicyForLogEntry))
+			ctx.Config().ApplyYAMLOrFail(t, ns, policies...)
+			defer ctx.Config().DeleteYAMLOrFail(t, ns, policies...)
+			t.Logf("Audit policy deployed to namespace %v", ns)
+
+			for _, cltInstance := range clt {
+				scopes.Framework.Infof("Validating Audit policy and Telemetry for Cluster %v", cltInstance.Config().Cluster.Name())
```
Comment (gargnupur) [inline tests/integration/telemetry/stackdriver/stackdriver_filter_audit_test.go:62]: nit: use t.Logf

File: pkg/test/framework/components/stackdriver/kube.go [hunk 24]
```diff
@@ -172,8 +191,14 @@ func (c *kubeComponent) ListLogEntries() ([]*loggingpb.LogEntry, error) {
...
+		if !strings.HasSuffix(l.LogName, logNameFilter) {
+			continue
+		}
 		// Remove fields that do not need verification
 		l.Timestamp = nil
+		l.Trace = ""
+		l.SpanId = ""
+		l.LogName = ""
```
Comment (gargnupur) [inline pkg/test/framework/components/stackdriver/kube.go:201]: why are we emptying this out? we should rather include this in logentry and use it in verification in json tmpl files.
This would help differentiate log entry files easily?

/cc @bianpengyuan

Comment (bianpengyuan) [reply pkg/test/framework/components/stackdriver/kube.go:201]: It is used to filter logs in the code above. It is filtered because it has project information in it, which relies on platform so not proper for verification.

**Review by gargnupur — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]
